jmxtrans-output/jmxtrans-output-cloudwatch/src/main/java/com/googlecode/jmxtrans/model/output/CloudWatchWriter.java

```diff
@@ -29,145 +29,187 @@
+	AmazonCloudWatch startAmazonCloudWatch() {
+		return AmazonCloudWatchClient.builder()
+				.withCredentials(InstanceProfileCredentialsProvider.getInstance())
+				.withRegion(Regions.getCurrentRegion().getName())
+				.build();
 	}
 
 	@Override
-	public OutputWriter create() {
-		return ResultTransformerOutputWriter.booleanToNumber(
-				booleanAsNumber,
-				new Writer(namespace, createCloudWatchClient(), createDimensions(dimensions))
-		);
+	public void close() throws LifecycleException {
+		if (amazonCloudWatch != null) {
+			amazonCloudWatch.shutdown();
+			logger.debug("Stopped CloudWatch client: {}", amazonCloudWatch);
+		}
+
+		super.close();
 	}
 
-	public static class Writer extends OutputWriterAdapter {
+	@Override
+	protected void internalWrite(Server server, Query query, ImmutableList<Result> results) {
+		PutMetricDataRequest request = new PutMetricDataRequest()
+				.withNamespace(namespace)
+				.withMetricData(toMetrics(results));
+		logger.debug("CloudWatch - Request: {}", request);
+
+		PutMetricDataResult response = amazonCloudWatch.putMetricData(request);
+		logger.debug("CloudWatch - Response: {}", response);
+	}
 
-		@Nonnull private final String namespace;
-		@Nonnull private final AmazonCloudWatch cloudWatchClient;
+	private ImmutableList<MetricDatum> toMetrics(ImmutableList<Result> results) {
+		return FluentIterable.from(results)
+				.filter(new Predicate<Result>() {
+					@Override
+					public boolean apply(Result result) {
+						return isNumeric(result.getValue());
+					}
+				})
+				.transform(new Function<Result, MetricDatum>() {
+					@Override
+					public MetricDatum apply(Result result) {
+						return toMetric(result);
+					}
+				})
+				.toList();
+	}
 
-		@Nonnull private final ObjectToDouble toDoubleConverter = new ObjectToDouble();
-		@Nonnull private final ImmutableCollection<Dimension> dimensions;
+	private boolean isNumeric(Object value) {
+		return NumberUtils.isNumeric(value);
+	}
 
-		public Writer(@Nonnull String namespace, @Nonnull AmazonCloudWatch cloudWatchClient, @Nonnull ImmutableCollection<Dimension> dimensions) {
-			this.namespace = namespace;
-			this.cloudWatchClient = cloudWatchClient;
-			this.dimensions = dimensions;
-		}
+	private MetricDatum toMetric(Result result) {
+		ImmutableList<String> typeNames = getTypeNames();
+		Map<String, String> typeNameValues = result.getTypeNameMap();
 
-		@Override
-		public void doWrite(Server server, Query query, Iterable<Result> results) throws Exception {
-			PutMetricDataRequest metricDataRequest = new PutMetricDataRequest();
-			metricDataRequest.setNamespace(namespace);
-			List<MetricDatum> metricDatumList = new ArrayList<>();
+		ImmutableList<Dimension> dimensions = FluentIterable.from(this.dimensions)
+				.append(toDimensions(typeNames, typeNameValues))
+				.toList();
 
-			// Iterating through the list of query results
+		return new MetricDatum()
+				.withDimensions(dimensions)
+				.withMetricName(result.getAttributeName())
```

Comment:
You should probably handle the case where result has a path (Heap memory attribute has a commited/used/init path for example)

Reply:
I completely missed this regression, thanks for pointing it out.